### DISCUSSION TITLE
Workaround for single-use auth tokens

### DISCUSSION THREAD

**Discussion by kylerummens** [supabase/supabase#7791] [Questions]
Some of the users of my web app are part of an organization that uses an enterprise email system (I think it might be Microsoft Authenticator?). Any time users are sent a reset-password email, the enterprise system opens the email and "clicks" on each of the links in the email to check for security risks. By the time the email shows up in the user's inbox, the reset-password link has already been used and is no longer valid.

I know that for security reasons (as shown in [this discussion](https://github.com/supabase/supabase/discussions/6004)) it isn't good practice to have a multi-use auth token.

Is there a suggested workaround for this scenario?

**Comment by GaryAustin1** [comment]
I'm using the newer {{.Token}} in email with the OTP the user just copies into your landing page and you send up for verification.  
https://github.com/supabase/supabase/discussions/5916
You might have a way to offer both options if easy to tell user when the link won't work...

Edit:  This has been discussed here in discussion and in issues, and there might be some other workaround, I just don't recall without searching.

**Reply by kylerummens** [reply]
I'm currently using my own email templates and generating the link using `supabase.auth.api.generateLink`.

Would the setup be different if I'm not using OTPs?

**Reply by GaryAustin1** [reply]
Sorry, I don't use generateLink...
There was this a few days ago, might help track down info.
https://github.com/supabase/gotrue/pull/537

**Reply by kylerummens** [reply]
What is the best method to get support for this issue?

**Comment by GaryAustin1** [comment]
I could swear I've seen a discussion or two (beside just use an otp code) on this in github or discord, but the terms are so general I can't find anything.

You might just want to open a new issue describing the problem with email server using the token, someone might link it to one of those cases involving the same thing, or someone from SB might respond with their thoughts.

**Reply by kylerummens** [reply]
Thanks, will do

**Comment by bkan36** [comment]
I've encountered this kind of issue with reset-password workflow 

my verify OTP return **Email link is invalid or has expired** every time.

I thought the problem was from supabase or SMTP and I fixed it by calling my function at the **mounted step** of my component
I use nuxt 3 (3.11.2)

here's my code

```
onMounted(async () => {
        await getUserFromToken()
    })

    async function getUserFromToken() {
        try {
            const tokenHash = route.query.th
            const type = route.query.type

            if (tokenHash && type) {
                const { data, error } = await supabase.auth.verifyOtp({
                    token_hash: tokenHash,
                    type,

                })
                if (error) throw error
                console.log(data)

                email.value = data.user.email

                console.log(email.value)
            }

        } catch (error) {
            console.error(error)
        }
    }
    
```

on supabase dashboard:
Authentication -> Email templates -> reset password

`<p><a href="{{ .SiteURL }}/reset-pwd?th={{ .TokenHash }}&type=recovery">Reset Password</a></p>`


if that's any help.
